### ISSUE TITLE
Please make it possible to compile on rust stable

### ISSUE THREAD

**Issue by PhilipDaniels** [peterjoel/needle#12]
Gives error 

error[E0554]: #![feature] may not be used on the stable release channel

 --> C:\Users\pdaniels\.cargo\registry\src\github.com-1ecc6299db9ec823\needle-0.1.1\src\lib.rs:2:1
  |
2 | #![feature(test)]
  | ^^^^^^^^^^^^^^^^^

error: aborting due to previous error

error: Could not compile `needle`.

**Comment by llogiq** [comment]
This is but one part of the problem (and the smaller part). This crate uses specialization, which isn't in stable yet, and may possibly not be for a while longer until all soundness holes have been ironed out.

That said, I believe it should be possible to make the use of specialization optional, and I'll send a PR to fix the benchmark inclusion in `src/` soon.

**Comment by PhilipDaniels** [comment]
@llogiq  I see that your PR has been merged. When I clone this repo now it does compile for me on stable (with a plain `cargo build`), even with the specialization still in - but when I try to compile your fork it complains about `#![feature(specialization)]`. I am very confused.

On the reddit thread I raised this on, another user @archer884 linked to his fork https://github.com/archer884/needle/commit/94bc179278e1f281d591348e06fc1df8e4d0f5c9 

where the specialization was removed simply by deleting the flag and removing `default` from the functions. Is this a valid technique?

**Comment by archer884** [comment]
@PhilipDaniels, the reason that works is that nothing within this particular crate is actually doing specialization. I guess that it could cause problems for other crates that want to use that feature (pure speculation), but I genuinely have no clue what this crate does or how it is used; I was just curious to see someone using specialization in the wild and noticed that, at least within this context, it's really not getting used.

Disclaimer: it's been awhile since I implemented anything using specialization...

But, basically, specialization is needed when there is an implementation of X in a given trait, but you don't want to use that implementation--you want to write your own. In that case, you need to specialize, which is to say you must override the original implementation. In this case, the function in question *has no implementation.* It's `skip_offset`, in `skip_search.rs`, on line 5.

https://github.com/archer884/needle/blob/master/src/skip_search.rs#L5

**Comment by PhilipDaniels** [comment]
@archer884 does that explain why when I update my Cargo.toml to point to your repo I get 

```
   | |_^ `[u8]` does not have a constant size known at compile-time
   |
   = help: the trait `std::marker::Sized` is not implemented for `[u8]`
   = note: required by `needle::BoyerMoore`
```

**Comment by archer884** [comment]
@PhilipDaniels, no, that doesn't make any sense to me. Could be unrelated or it could be some unintended effect I just don't grok. It's telling you that the trait requires an object with a size and that an unsized array doesn't have a size. Which... You know. Not helpful, I guess?

I looked to see if you're messing with this on any of your public repositories, but that doesn't seem to be the case. Maybe you could link to a gist containing an example of the problem?

**Comment by PhilipDaniels** [comment]
@archer884 The code is designed to process some log files that we produce at work so I have never pushed it publicly. I've been hacking around with 5 or 6 different ways to do it so it's not very clear. I will try and make a minimal repro, but it won't be tonight.

I have had some strange compilation problems while trying to point to repos on Github rather than the official version on crates.io. I suspect they will all go away if Peter publishes a new version to crates.io without specialization :-)

**Comment by PhilipDaniels** [comment]
@archer884 I made a very small program to demonstrate https://github.com/PhilipDaniels/needle-harness  
Didn't need to worry about searching log files or anything, we can just show the issues directly in main(). There are comments in the Cargo.toml describing various combinations. The code as checked in is what I expect to work based on the docs. It does run on nightly.

**Comment by archer884** [comment]
Hi, @PhilipDaniels, thanks for doing all the legwork on this. What it's saying is just that you need to import the `SearchIn` trait. For whatever reason (and I can't guess what without digging deeper--unfortunately I'm pretty swamped right now), this trait is already imported (?) or not required by the original version using specialization.

Anyway, the code below runs on stable without specialization.

```rust
extern crate needle;

use needle::{BoyerMoore, SearchIn};

fn main() {
    let haystack = b"blah blah blah Something=Tasty blah blah blah";
    let something_finder = BoyerMoore::new(&b"Something="[..]);
    
    match something_finder.find_first_in(haystack) {
        None => println!("No match"),
        Some(i) => println!("Matched at {}", i)
    };
}
```

I'm guessing your error from the other day was a missing `&` sigil. Typos are a pain when trying to debug unrelated crap.

(Edited after I realized that you had already noticed the different error message.)

P.S.

The above code is using this fork:

```toml
needle = { git = "https://github.com/archer884/needle" }
```

**Comment by peterjoel** [comment]
Thanks for all this. I think the reason that Specialization was required was to support searching in generic slices, while optimising for  `u8`. The speed difference is so significant, that I actually wasn't sure if I'd continue to support the generic case.

**Comment by PhilipDaniels** [comment]
Thanks @archer884 for your help, my little program is now compiling on stable with the aid of your fork.

@peterjoel Certainly I would expect the vast majority of usages to be searching in byte arrays/strings so you would have my support for removing the specialization if it meant there was a version on stable. Your lib is the only implementation of Boyer-Moore on Cargo so I think it definitely helps round out the ecosystem.
